### DISCUSSION TITLE
CORS Issue / Preflight Request using PocketBase

### DISCUSSION THREAD

**Discussion by jspann93** [pocketbase/pocketbase#2136] [Q&A]
Hey all! I’m working on a soccer lineup builder application that uses PocketBase to keep track of player submissions. I’m almost done with it and started testing from another computer to make sure other uses can update the database. That’s when I noticed this issue.

When I load the page from a computer other than my own, these errors pop-up in the console.
![image](https://user-images.githubusercontent.com/111022851/227025449-8f15833f-1ca5-4a3b-a5d7-ed849f4b5510.png)

From my own computer, I get a warning that I’m requesting a resource that I can only access because of my network position.
![image](https://user-images.githubusercontent.com/111022851/227025493-470bab4a-5625-4f31-bedb-f94f0dc74d97.png)

It sounds like I need to update something with either PocketBase or [fly.io](http://fly.io/) to allow other uses to access the database. Does anyone have any experience with this? I currently don’t have any sign-in methods on my PocketBase database which I thought would allow everyone to have access. Any help would be greatly appreciated. Thanks!

The webpage is here: https://us-soccer-lineup-builder.fly.dev/
The code is here: [GitHub - jspann93/US-Soccer-Lineup-Builder](https://github.com/jspann93/US-Soccer-Lineup-Builder)

**Comment by ganigeorgiev** [comment]
Hm, I don't this is caused by the PocketBase server because by default all origins are allowed (aka. `*`), unless you've specified the `--origins` flag.

Based on https://github.com/jspann93/US-Soccer-Lineup-Builder/blob/main/pb_data/Dockerfile it seems that you are serving the frontend with some custom http server, not with PocketBase so that could be the cause of the issue..

Additionally, please note that your repo is structured wrong. **`pb_data` shouldn't be public**, otherwise you are risking someone to download your entire database - https://github.com/jspann93/US-Soccer-Lineup-Builder/tree/main/pb_data (it is also accessable via the url; make sure to fix it if it is not intentional!).

If you want PocketBase to serve static files, you can place them inside `pb_public`.

**Reply by jspann93** [reply]
Yes, I'm using fly.io to host the frontend using the instructions here: https://github.com/pocketbase/pocketbase/discussions/537
Could it be something in my dockerfile or fly.toml causing the issue? Where does the --origins flag you mentioned get set?

Also, none of the data is in the database is particularly sensitive, but I will move pb_data to gitignore just to follow best practices. Thanks for the help!

**Reply by ganigeorgiev** [reply]
> Yes, I'm using fly.io to host the frontend using the instructions here:

Ok, but in [your Dockerfile](https://github.com/jspann93/US-Soccer-Lineup-Builder/blob/main/pb_data/Dockerfile) `pierrezemb/gostatic` image is used and it seems that you are serving the static files with it, not with PocketBase.

>  Where does the --origins flag you mentioned get set?

It is part of the `serve` initialization arguments. You don't have to set it explicitly. By default all origins (`*`) are allowed.

> Also, none of the data is in the database is particularly sensitive, but I will move pb_data to gitignore just to follow best practices.

You should exclude it from the git repo but most importantly **nothing from `pb_data` should be public accessible!**. The only public HTTP accessible directory must be your static files (if you are serving the static files with PocketBase then that could be `pb_public`).

**Reply by jspann93** [reply]
Hmm I'm not sure what I did there. I changed my dockerfile to match what was in [the guide](https://github.com/pocketbase/pocketbase/discussions/537) and rebuilt the app. Now the page doesn't load after deployment with an http 503 error. 

<img width="1350" alt="image" src="https://user-images.githubusercontent.com/111022851/227048181-b8aae291-4750-4fe6-9f7d-4385420abfac.png">

**Reply by jspann93** [reply]
Clearning browser cache and refreshing the page gets me:
{"code":404,"message":"Not Found.","data":{}}

**Reply by ganigeorgiev** [reply]
The mentioned guide takes care for deploying only the PocketBase server.

It is up to you to deploy your static files and `pb_public` dir. This could be done several ways:
- use the `ADD` or `COPY` Dockerfile instruction to copy you local `pb_public` into the image
- or mount the static file directory as another volume (similar to how you register the `pb_data` in the fly.toml)
- or wrap both `pb_data` and `pb_public` in for example a single `pb_volume` directory and use that for the volume (in the Dockerfile you must specify explicity `--dir` and `--publicDir` location pointing to `/pb/pb_volume/pb_data` and `/pb/pb_volume/pb_public`)

**Reply by jspann93** [reply]
Sorry for asking so many silly questions. I have almost zero idea what I'm doing lol. 

So, to clarify, your recommendation is to move my static files (index.html, lineup.js, and lineup.css) to a directory called pb_public. Then, edit the dockerfile to copy pb_public into the image with ADD or COPY. Is that correct?

**Reply by ganigeorgiev** [reply]
Yes. The idea is to have your static files inside `pb_public` and that directory to be available in the same directory as the executable.

You can find some information about `ADD` and `COPY` in https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy, but in general you should be able to add something like the below right before the last `CMD` call:

```dockerfile
COPY ./pb_public /pb/pb_public/
```

**Reply by jspann93** [reply]
Okay, so is this the correct file configuration:

At the top-level here I have my pb_data, pb_public, and pb_migrations, as well as the dockerfile and pocketbase executable. 
<img width="757" alt="image" src="https://user-images.githubusercontent.com/111022851/227057241-07c0b180-29bb-42a4-b96e-2d8a7a411c53.png">

pb_public now contains my static files. 
<img width="755" alt="image" src="https://user-images.githubusercontent.com/111022851/227057457-d50528ec-3112-4097-9318-6b2df9dfca57.png">

pb_data has the database and fly.toml
<img width="771" alt="image" src="https://user-images.githubusercontent.com/111022851/227057521-1f3e18a7-462c-4b2d-a1ea-e6c460ce12c7.png">

**Reply by ganigeorgiev** [reply]
Your local `pb_data` doesn't really matter in this case. When you deploy it on fly.io, it will use the `pb_data` volume that you've created.

Note that if you have `pb_migrations` you can add them also to the Dockerfile similar to `pb_public`.

**Comment by jspann93** [comment]
Okay, so I've modified my [dockerfile](https://github.com/jspann93/US-Soccer-Lineup-Builder/blob/main/Dockerfile) to serve the front end with PocketBase. When I redeployed, [the site](https://us-soccer-lineups.fly.dev/) came up and functioned properly as before, but I still run into some type of CORS issue. 

<img width="762" alt="image" src="https://user-images.githubusercontent.com/111022851/227215264-12a20a75-9a2e-49ea-ac76-6e526785ea33.png">

Also, perhaps unrelated, I had to run ./pocketbase serve from the directory before I could interact with the database. I thought the dockerfile would do that for me but I suppose it didn't. 
<img width="1060" alt="image" src="https://user-images.githubusercontent.com/111022851/227216120-f363f0a0-1160-4f6b-a891-088427e8b742.png">

**Reply by ganigeorgiev** [reply]
I don't see any CORS issues, but please note that you have to also update the SDK url in https://github.com/jspann93/US-Soccer-Lineup-Builder/blob/main/pb_public/lineup.js#L20.

**Reply by jspann93** [reply]
The CORS issue doesn't pop up when I view the webpage because I have privileged network position as the admin, but if I try to use the webpage from a different device, the submit button doesn't work properly and this CORS issue pops up upon inspect.
![image](https://user-images.githubusercontent.com/111022851/227220752-a189a137-8beb-4152-8526-ef1a9cca6452.png)


I don't quite follow regarding the SDK url. I took that straight from the [JS SDK instructions](https://github.com/pocketbase/js-sdk). What should it be instead?
<img width="488" alt="image" src="https://user-images.githubusercontent.com/111022851/227219611-be30b37f-24f3-4743-a13f-381278d39842.png">

**Reply by ganigeorgiev** [reply]
> I don't quite follow regarding the SDK url. I took that straight from the [JS SDK instructions](https://github.com/pocketbase/js-sdk). What should it be instead?

`http://127.0.0.1:8090` is your localhost address. You need to replace it with the PocketBase public url, aka. `https://us-soccer-lineup-builder.fly.dev/`.

I don't think your issue is related to PocketBase.

If you are making browser requests to 3rd party websites it is up to the 3rd party server whether they allow your origin or not. Also, at least from the logs `https://trycom.s3.amazonaws.com/scrapper-sdk.js` doesn't seem to be publicly accessible.

**Reply by jspann93** [reply]
Okay, you've lost me a tad with the PocketBase public url. I changed the URL from http://127.0.0.1:8090 to https://us-soccer-lineups.fly.dev/, but now I get a 404 error when trying to access the database. 
https://us-soccer-lineups.fly.dev/ is the url users will use to interact with the webpage, but the data collected from that page needs to be sent to the pocketbase collection so I can see it, right? What is the url for that pocketbase collection if the localhost address is no longer being used?

<img width="1072" alt="image" src="https://user-images.githubusercontent.com/111022851/227231487-ca48ba53-6210-4dd2-b020-9d8194851b68.png">

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't really know how to explain it better.

When making request client-side (aka. with the browser), the localhost address refers to the client 
local env. You need to provide the public PocketBase url so that when a user open your website he can send request to the remote PocketBase server.

**Reply by jspann93** [reply]
Right, I understand that conceptually. I just assumed the website would have a different url than the pocketbase server. I want users to be able to access the site and use its interface to modify data in the database, but they shouldn't be able to see the database itself. How do I see the database once it's created?

Sorry for all the confusion.

**Reply by ganigeorgiev** [reply]
>  How do I see the database once it's created?

I don't understand what do you mean. If you are asking for the Admin UI, you should be able to access it via https://us-soccer-lineups.fly.dev/_/ (aka. replace `https://127.0.0.1:8090` with your public url).

**Reply by jspann93** [reply]
Thank you so much. I'm not sure why that wasn't clicking for me, but I understand now. I have rebuilt everything and I think it's about 99% there. 

I have one final question I'm hoping you can help me with, then I'll leave you alone :)

I can successfully add records to each collection using the webpage interface, but for whatever reason, incrementing my count field doesn't seem to work. I could do it when I used the localhost database, but now using the public url, nothing happens (code didn't change). I made sure the API rules allow for all users to view, create, and update records. 
<img width="665" alt="image" src="https://user-images.githubusercontent.com/111022851/227248217-635584d1-b3bd-48d9-8307-eab80ef8be1d.png">

Do you see anything wrong with the syntax here?
<img width="988" alt="image" src="https://user-images.githubusercontent.com/111022851/227248772-610cc60f-7037-41d4-8eca-7d4538d6d4f2.png">

**Reply by ganigeorgiev** [reply]
In the code screenshot, the update call is for `men_formation_selections`, while the Admin UI screenshot shows `men_player_selections`.

If that's not the issue, you can try to wrap the `await` calls with `try/catch` and see if there any errors. Alternatively, you can also check the request logs in the "Admin UI > Logs".

**Reply by jspann93** [reply]
Yeah those are two separate collections but I'm trying to update the count for both and neither is working. 

I've been looking at the logs and all of the show status 200 which is good. 
<img width="1168" alt="image" src="https://user-images.githubusercontent.com/111022851/227255200-ac9eb6de-6b60-4276-bbb9-f8bfea61daa6.png">

I'll try wrapping the await calls in try/catch to see if any errors pop up. I stepped through the code with breakpoints and everything looks fine so I'm not sure what's going on.

**Reply by ganigeorgiev** [reply]
Are you sure that `count` is a `number` collection field?

**Reply by jspann93** [reply]
No errors were caught with try/catch either...

**Reply by jspann93** [reply]
Yes, count is definitely a number field. 
<img width="793" alt="image" src="https://user-images.githubusercontent.com/111022851/227257082-05edc6d8-d87a-4020-8372-a8684f41d558.png">

<img width="679" alt="image" src="https://user-images.githubusercontent.com/111022851/227257217-b607a814-9042-4e6d-b2b7-1e75e1030165.png">

**Reply by jspann93** [reply]
I tweaked the code to change count to 50 instead of incrementing it. That worked. So there's something with the increment syntax that isn't working.

**Reply by jspann93** [reply]
I don't know exactly why but it didn't like the  "count+": 1 syntax. When I grabbed the record, copied the count locally, incremented the count locally, then stored the new value in count, it worked fine.

**Reply by ganigeorgiev** [reply]
I've just tested it on the live demo and data like `{"count+": 1}` works fine for me. We also have integration tests that verify the correct behavior.

I'm sorry but I'm getting tired of guessing what you are doing. If you think that there is an issue with PocketBase, please provide reproducible steps and I'll investigate it.

**Reply by jspann93** [reply]
I'm sorry for the trouble, and I thank you for all your time.
